Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
+        diagnosticsInSourceLine +=
+          diagnostics
+          .flatMap(\.notes)
+          .filter { note in
+            note.location(converter: sourceLocationConverter).line == sourceLineNumber
+          }
+          .map(FlattenDiagnostic.note)
 
-    // First, we need to put each line and its diagnostics together
-    var annotatedSourceLines = [AnnotatedSourceLine]()
+        let suffixText =
+          suffixTexts
+          .compactMap { (position, text) in
+            sourceLocationConverter.location(for: position).line == sourceLineNumber ? text : nil
+          }
+          .joined()
 
-    for (sourceLineIndex, sourceLine) in slc.sourceLines.enumerated() {
-      let diagsForLine = diags.filter { diag in
-        return diag.location(converter: slc).line == (sourceLineIndex + 1)
+        return AnnotatedSourceLine(
+          diagnostics: diagnosticsInSourceLine,
+          sourceString: sourceLine,
+          sourceLineNumber: sourceLineNumber,
+          suffixText: suffixText
+        )
       }
-      let suffixText = suffixTexts.compactMap { (position, text) in
-        if slc.location(for: position).line == (sourceLineIndex + 1) {
-          return text
+  }
+
+  /// Computes line number ranges that should be printed, based on the presence of annotations and the selected context.
+  ///
+  /// This function scans an array of `AnnotatedSourceLine` to determine which lines are
+  /// annotated with diagnostics or notes. Depending on the `context` property, this method behaves as follows:
+  ///
+  /// - `.limited(n)`: For each annotated line, a `Range<Int>` is created to
+  ///   indicate the span of lines around the annotated line that should also be printed, limited to `n` lines before and after.
+  ///
+  /// - `.full`: A single range spanning all the lines will be returned, effectively printing the entire source code alongside diagnostics.
+  ///
+  /// - Parameters:
+  ///   - annotatedSourceLines: An array of `AnnotatedSourceLine`, each of which represents a line
+  ///     of source code and any annotations associated with it.
+  ///
+  /// - Returns: An array of `Range<Int>`, each specifying a range of line numbers that should be
+  ///     printed due to the presence of annotations or the context setting.
+  private func computeLinesToPrintRanges(from annotatedSourceLines: [AnnotatedSourceLine]) -> [Range<Int>] {
+    switch contextRange {
+    case .limited(let contextSizeLimit):
+      return
+        annotatedSourceLines
+        .compactMap { annotatedLine -> Range<Int>? in
+          let lineNumber = annotatedLine.sourceLineNumber
+          if annotatedLine.isFreeOfAnnotations {
+            return nil
+          } else {
+            return Range<Int>(uncheckedBounds: (lower: lineNumber - contextSizeLimit, upper: lineNumber + contextSizeLimit + 1))
+          }
         }
+    case .full:
+      return [Range<Int>(uncheckedBounds: (lower: 0, upper: annotatedSourceLines.count))]
```

Comment:
Same here. Let’s not use the unchecked initializer. It’s also easier to read.

```suggestion
      return [0..<annotatedSourceLines.count]
```